### DISCUSSION TITLE
[discussion / planning] How to move forward with full ONNX support

### DISCUSSION THREAD

**Discussion by felixdittrich92** [mindee/doctr#981] [General]
Hi @frgfm @charlesmindee and all other :wave:,

we will be soon in an state with the library where we are able to export the models into onnx format. (full support planned 0.6)
So i think it would be good to discuss some further work in this topic.

As a user i want:

- export models into onnx to use it into a production system
- load the exported models directly in the library like `ocr_predictor` without installing PyTorch or Tensorflow as dependency onnxruntime should be enough
- be able to configure the runntime (for example CPUExecutionProvider / GPUExecutionProvider)

There are some open question:

- How want we to design the loading ? I think we have to encapsulate it from the default `ocr_predictor` maybe introducing a `onnx_ocr_predictor` makes sense ? 
- We need to do the post processing afterwards so any ideas how we can adopt this ?
- We need to store some information about the onnx exported models for example which architecture, vocab, and so on .. maybe working with config files in the same way we have done this for huggingface `push_to_hub` and `from_hub` ?

Feel free to add some additional to this list if i missed anything :hugs: 

Any feedback is very very welcome :heart:

**Comment by charlesmindee** [comment]
Hi Felix,
Thanks for bringing this on the table, I like the idea of having a `onnx_ocr_predictor` or something equivalent that won't require tensorflow nor torch. For the postprocessing, it would be great to share it with the tensorflow/torch models since it shouldn't be DL-framework related, WDYT ?

**Reply by felixdittrich92** [reply]
Hi Charles,

yes i have had the same in mind.
We have to take care not to run into circular imports and that we do not need any framework specific code (need an extra setup build for onnxruntime)
On the other hand we need anything to store the information about the model to load (which postprocessor / vocab / etc.) 
I will check if we can store it inside the exported model but i think we need to save config files while exporting which provides the information for example export creates a directory which contains the .onnx and a .json (like hf stuff) or do you have any other idea ? :)

**Reply by frgfm** [reply]
I agree with Charles, I think that will lead us to add ONNX as a supported DL backend 

But that means that from now on, we'll have different support status for features depending on the backend: training & co won't be available in ONNX. This is entirely doable (cf. #984), but I suggest first focusing on: export capacity with docTR, loading this is minimal environment.

Fortunately I have experience in this, for other projects, I have uploaded ONNX models on HF hub, and made a demo or API that was using only ONNX runtime + PIL + numpy for image loading, preprocessing, inference and post processing. The only difference I see here is that we need OpenCV and other deps for some model's postprocessing :thinking: 

If we step back to assess the needs from a user perspective :
- from a loaded TF/pytorch model, being able to export to ONNX file
- performing inference with that ONNX file (the question is about the situation in which they would do so : production backend, development?)

Considering the whole purpose of ONNX, I think the production / backend situation is the priority but I might be biased. What do you think? :)

**Reply by felixdittrich92** [reply]
:wave: 
Same for me also with native mobile Onnxruntime deployment outside from Python :)
The development scenario sounds for me like ORT training that's what you mean with development? If so we should focus on prod our current training workflow works good an i see this feature addition as part to make docTR also usable in a prod system

**Reply by frgfm** [reply]
for development I mean people using interactive shell, or loading the model to write other pieces of code (not only to perform inference)

I agree about the rest :)

**Reply by felixdittrich92** [reply]
ok so i think we have a fix main topic/feature for v0.6.0 :smile:

**Reply by frgfm** [reply]
The tricky part will be porting the post-processing to a numpy or ONNX compatible operation :sweat_smile:

**Reply by felixdittrich92** [reply]
I think we do not need this why not adopting it ? Yes in this case the onnx export would be bound to use it inside docTR wdyt ?

**Reply by frgfm** [reply]
Alright, but docTR, as is, is not light. Meaning that it's not exactly optimal for production.
Even without the DL backend like Pytorch or TF, docTR comes with a lot of dependencies that might not be relevant for production situations :man_shrugging: 

But either way, this is a very large topic: balance of docTR between dev / user types, dev / production situations

**Reply by felixdittrich92** [reply]
Yes i agree, i would say in a first step it would be good to start with a PoC after the 0.5.2 release and the TF mobilenet arch fix. 
If this is done we can further check if it makes sense/ is possible to provide maybe a minimal build. wdyt ? 

(And yes, it would be easiest if we managed to make the postprocessing onnx compatible, then everything else would be much easier and more flexible. But as you said, there are quite a few places that can break our necks :face_with_head_bandage: :sweat_smile:)

On the other hand in this case docTR would be the first library which provides End-to-End OCR in an full Onnxruntime environment (or do you know anyone who can provide something ?) which would definitely be a blatant unique point

**Reply by frgfm** [reply]
We got a few options, the easiest one for us would be to have a docTR core inference build: ONNX centered, no dev features, only inference (I.e. for production)

So for instance, the post processing can be left to the user to perform and we only provide the logits (not optimal I agree)
Going for full ONNX compatibility is going to be very long

My point is that:
- docTR can natively provide ONNX export
- but providing ONNX inference feature without discarding dev features will make any production docker between 3 and 5 times larger than what it needs (mostly because of Pytorch & TensorFlow)
- so I suggest considering the ONNX version something to implement for a production situation and the best way I can see for this: implement it in the API boilerplate that we provide. I've already done this for personal projects, it's blazing fast, so much lighter. The only trick is that I don't have postprocessing that is this complex.

**Reply by felixdittrich92** [reply]
@frgfm yes but in this case we ignore the biggest advantage to get an well structured output which is easy to use . I think if we find a solution for the pre- and post-processing it is also possible to have a lightweight build only for inference also inside docTR. I will test something after we are done with 0.5.2 and i'm back from vacation 😊

**Reply by frgfm** [reply]
Yup, ideally, if we could redesign everything instantly:
- most independent modules should be set as extras
- DL backend: for dev & training, PyTorch and TF should be available. for inference, also have ONNX
- post-processing: everything should be implemented in numpy, PyTorch or TF (for the last two, that can be passed to ONNX then)

But we need to find a middleground here :)

**Comment by ajkdrag** [comment]
@felixdittrich92 I see the docs is updated to export to Onnx, but is the inference/loading working as of now?

**Reply by felixdittrich92** [reply]
Hey @ajkdrag 👋,

We don't have a inference pipeline with onnxruntime atm.
With the multiframework approach, this is not so easy to implement. However, we are currently discussing to what extent it makes sense to keep TF instead of Onnxruntime (Pipeline) & PT, as the upgrade to Keras v3 with TF >= 2.16 would require a massive effort.

CC @odulcy-mindee @frgfm @charlesmindee

**Reply by frgfm** [reply]
Hey there!
Thanks, I remember when we made that call: the bottom line is that ONNX is more one-sided than the other DL backends (inference-only). So in my perspective, if we had infinite resources, we'd make it so that the entire predictor can be exported to ONNX

And after that, we can create a notebook but users should be implementing it, and we should not fragment the core library too much :)

Back to your question @ajkdrag, right now, you'd have to assemble several ONNX compatible parts and convert the code in-between. If we were to make that jump, my suggestion is to focus on making all pre/post-processing using PyTorch which we'll be able to convert to ONNX at will.

I personally don't have enough bandwidth to take care of that, but it could be an interesting challenge to split this into subtasks where the community could help :v: 

Cheers!

**Reply by ajkdrag** [reply]
Only post-processing is missing, right? The preprocessing is already in pytorch i believe.

**Reply by felixT2K** [reply]
@ajkdrag pre-processing should not be the problem. But post-processing is tricky because especially for the detection we use some `shapely` and `opencv` code

**Reply by nospotfer** [reply]
Do we really need to convert the post-processing pipeline to ONNX as well? As I user, I'd be more than happy to have a predictor or even different .onnx models (detection.onnx, recognition.onnx, etc) and leave the post-processing as methods depending on shapely, opencv and any other libraries.

**Reply by felixdittrich92** [reply]
@nospotfer Good time to ask :D

See: https://github.com/felixdittrich92/OnnxTR (I have some problems with the gpu installation atm but the cpu version works well already)

**Reply by felixdittrich92** [reply]
@odulcy-mindee Maybe we can discuss later on if we want to move the repo to the mindee org otherwise keeping it on my private account would also be fine - it's much better to keep this seperated instead of having a third backend in doctr i think
(I will take care to keep it in sync with doctr)

**Reply by nospotfer** [reply]
Wow! Perfect timing!
I'll have a look. GPU inference is a nice to have, but I understand most of the people (including myself) are more than happy with a CPU inference pipeline.

**Comment by felixdittrich92** [comment]
I think we can close the discussion now with https://github.com/felixdittrich92/OnnxTR :)
